Add helm chart

stevehipwell opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

Adds a Helm chart based on the official manifests in the repo; the Helm repo will be hosted by GitHub pages at https://kubernetes-sigs.github.io/external-dns/.

Adds required GitHub Actions CI.

- A PR making changes to files in `/charts/external-dns/**` will trigger the lint and test action
  - If the chart has changed then the chart version should be updated accordingly or else the CI will error
- If changes are made to files in `/charts/external-dns/**` on the master branch a release will be made
  - The `gh-pages` branch needs to be setup so the release action can use it as the Helm repo

The chart is fully documented but I've not linked it in to any of the other documentation as maybe this should be a soft launch? Otherwise could someone point me in the right direction to put the documentation in the correct place.

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #1941

**Checklist**

- [X] Unit tests updated
- [X] End user documentation updated

stevehipwell commented:
/assign @Raffo

dirien commented on charts/external-dns/README.md:
````diff
@@ -0,0 +1,66 @@
+After you've installed the repo you can install the chart.
+
+```shell
+helm upgrade --install external-dns/external-dns
````
Would be helpful to have an example for another provdider like Hetzner or DigitalOcean.

They expect the Token via ENV.

```yaml
 env:
        - name: HETZNER_TOKEN
          value: "YOUR_HETZNER_DNS_API_KEY"
```

So to have it `GitOps` conform you should set the env like this:

```yaml
env:
            - name: DO_TOKEN
              valueFrom:
                secretKeyRef:
                  name: <NAME>
                  key: <KEY_NAME>
```
And should define a secretref in the values.yaml

stevehipwell replied:
The current chart isn't actually configured to work with AWS, it just uses AWS as a valid argument so that it can be deployed to test the Kubernetes resources, there are no credentials. To use any provider the correct credentials must be provided, but this should be covered in the provider specific documentation and isn't specific to the Helm chart. For example AWS could be configured by an env mounted secret as seen in your DO example above, it could also be configured as a volume mounted secret, however it should be configured as a service account annotation using IRSA.

dirien replied:
i did not spot an option to set a custom env variable and mount and ref to it. Maybe i missed the place in the template code. I just need a way to support providers with API Tokens without havin to patch the external-dns deplyoment.

stevehipwell replied:
You've got the `env` value which is in _values.yaml_ and documented in the chart README.

dirien replied:
ok i must missed this 😅 .

But i would add an example in the Readme, on how to use it that way. I would add it but i cant commit in your branch.

Otherwhise, i wait it gets merged and make a new PR. 

It would help to increase the DX and accessibility

stevehipwell replied:
It's pretty idiomatic to well authored Helm charts, I'm happy to add a comment in the README values section but I don't think an example is appropriate here. If we start adding example for using idiomatic Helm patterns, especially ones which just capture the Kubernetes objects and pass them through as is the case here, we'd need an example of all the areas like this such as volumes and volume mounts etc and would be basically teaching people how to use Kubernetes & Helm.

dirien replied:
Yepp, some comment in the Readme and a hyperlink to the external-dsn main readme with the list of providers would be fine too! There its written how the provider expects som auth stuff.  Fine with this 😄 

Thanks for Steve for the discussion!

stevehipwell replied:
I've updated the README with links to the relevant K8s APIs that are supported in the _values.yaml_ and linked the `provider` value to the README so I'm going to mark this thread as resolved.

dirien reviewed (changes requested):
Hi, 

Left some comments.

stevehipwell commented on charts/external-dns/Chart.yaml:
```diff
@@ -0,0 +1,22 @@
+icon: https://github.com/kubernetes-sigs/external-dns/raw/master/img/external-dns.png
+sources:
+  - https://github.com/kubernetes-sigs/external-dns/
+  - https://github.com/stevehipwell/helm-charts/
```
Good spot, I'd intended to remove that line but must have forgotten.

njuettner commented:
Seems like the linter is failing 😞 

```
Events:
  Type     Reason            Age                 From               Message
  ----     ------            ----                ----               -------
  Warning  FailedScheduling  33s (x5 over 5m1s)  default-scheduler  0/1 nodes are available: 1 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate.
```

stevehipwell commented:
@Raffo could you re-run the lint-test action for the above mentioned reason?

Raffo commented:
@stevehipwell  I enabled CI, the linter is still unhappy.

stevehipwell commented:
@Raffo I can't get to the logs but was the Kind cluster creation good or did that never get healthy?

stevehipwell commented:
I got to the logs, it looks like it was Kind. Could you re run the action?

Raffo commented:
@stevehipwell kind being unkind uh 😅 rerunning!

Raffo commented:
And it failed again. I think the build deserves a deeper look on why this is happening.

jordiprats commented:
Hi guys,
I'm hoping this would merged so I took a look, it says:

https://github.com/kubernetes-sigs/external-dns/pull/2208/checks?check_run_id=3320498211

 0/1 nodes are available: 1 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate

I'm not familiar with the cluster were you are running this but sounds like it is spawning a new node to run this pod and you are not waiting long enough for the node to become ready.

stevehipwell commented:
If I remember correctly it's possible to increase the timeout on the Kind action. I run a number of charts with this exact config and up until recently hadn't seen this issue; up until this point I think I've only ever seen it twice in well over 50 runs.

k0da commented:
@stevehipwell you can use [act](https://github.com/nektos/act) to simulate GHA runs locally. Also we found k3d project is more lightweight to spawn k8s clusters for CI. [Here](https://github.com/AbsaOSS/k3d-action) is the action for k3d. JFYI

stevehipwell commented:
Thanks @k0da, as I said above this pattern had been working for years and hundreds of chart PRs so I've no idea why it's suddenly got brittle.

stevehipwell commented:
@Raffo I've updated the Kind action version (I thought I'd already done this) and added a longer wait for the cluster to be ready which should solve this.

Raffo reviewed (commented):
@stevehipwell CI is green and the chart content looks all right to me.  However, I think that we need to remove/change the part on releasing the chart: we have a consolidated release process described in https://github.com/kubernetes-sigs/external-dns/blob/master/docs/release.md and we would need to make creating the release part of it. I also would love to avoid having multiple github releases for what are code and helm changes. Would you be able to propose a change to incorporate that?

stevehipwell commented:
@Raffo unfortunately the chart releaser will create a GitHub release for each chart release and there is no way to change this behaviour. You'll see in the release action that the suffix _-helm-chart_ is added to distinguish the release.

Would it be easier to merge this PR and add the documentation separately as the first release only requires the _gh-pages_ branch to be created with GitHub pages enabled and the PR to be accepted?

onedr0p commented:
@stevehipwell would it be beneficial to include the external-dns `DNSEndpoint` CRD in a crds folder? See the [this chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) for an example. Otherwise I am not seeing how to install the `DNSEndpoint` CRD.

stevehipwell commented:
@onedr0p if the CRDs are fully supported then they should be in the CRDs chart folder, but as they're not in the standard manifest install I've not added them.

However if you read the Helm 3 docs you will see that although CRD support is partially implemented you shouldn't use it and disable installing CRDs with the `--skip-crds` flag then manage your CRDs with kubectl.

Raffo commented:
> Would it be easier to merge this PR and add the documentation separately as the first release only requires the gh-pages branch to be created with GitHub pages enabled and the PR to be accepted?

This sounds good to me.

Raffo commented:
/approve
/lgtm
